test/optimisation/Optimisation.jl

```diff
@@ -0,0 +1,557 @@
+            logp, vi = DynamicPPL.tilde_observe(context.context, right, left, vi)
+            return context.loglikelihood_weight, vi
+        end
+
+        @model function model1(x)
+            μ ~ Uniform(0, 2)
+            x ~ LogNormal(μ, 1)
+        end
+
+        @model function model2()
+            μ ~ Uniform(0, 2)
+            x ~ LogNormal(μ, 1)
+        end
+
+        x = 1.0
+        w = [1.0]
+
+        @testset "With ConditionContext" begin
+            m1 = model1(x)
+            m2 = model2() | (x=x,)
+            ctx = Turing.Optimisation.OptimizationContext(DynamicPPL.LikelihoodContext())
+            @test Turing.Optimisation.OptimLogDensity(m1, ctx)(w) ==
+                  Turing.Optimisation.OptimLogDensity(m2, ctx)(w)
+        end
+
+        @testset "With prefixes" begin
+            function prefix_μ(model)
+                return DynamicPPL.contextualize(
+                    model, DynamicPPL.PrefixContext{:inner}(model.context)
+                )
+            end
+            m1 = prefix_μ(model1(x))
+            m2 = prefix_μ(model2() | (var"inner.x"=x,))
+            ctx = Turing.Optimisation.OptimizationContext(DynamicPPL.LikelihoodContext())
+            @test Turing.Optimisation.OptimLogDensity(m1, ctx)(w) ==
+                  Turing.Optimisation.OptimLogDensity(m2, ctx)(w)
+        end
+
+        @testset "Weighted" begin
+            function override(model)
+                return DynamicPPL.contextualize(
+                    model,
+                    OverrideContext(model.context, 100, 1)
+                )
+            end
+            m1 = override(model1(x))
+            m2 = override(model2() | (x=x,))
+            ctx = Turing.Optimisation.OptimizationContext(DynamicPPL.DefaultContext())
+            @test Turing.Optimisation.OptimLogDensity(m1, ctx)(w) ==
+                  Turing.Optimisation.OptimLogDensity(m2, ctx)(w)
+        end
+    end
+
+    @numerical_testset "gdemo" begin
+        @testset "MLE" begin
+            Random.seed!(222)
+            true_value = [0.0625, 1.75]
+            true_logp = loglikelihood(gdemo_default, (s=true_value[1], m=true_value[2]))
+
+            function check_success(result)
+                optimum = result.values.array
+                @test all(isapprox.(optimum - true_value, 0.0, atol=0.01))
+                @test result.optim_result.retcode == SciMLBase.ReturnCode.Success
+                @test isapprox(result.lp, true_logp, atol=0.01)
+            end
+
+            m1 = Turing.Optimisation.estimate_mode(
+                gdemo_default, MLE()
+            )
+            m2 = maximum_likelihood(
+                gdemo_default, true_value, OptimizationOptimJL.LBFGS()
+            )
+            m3 = maximum_likelihood(
+                gdemo_default, OptimizationOptimJL.Newton()
+            )
+            # TODO(mhauru) How can we check that the adtype is actually AutoReverseDiff?
```

Comment:
Annoyingly, I think the "best" approach is just

``` julia
types_used = []

@model function demo()
    global types_used
    push!(types_used, eltype(__varinfo__[:]))

    # Do other stuff.
end
```

and then check this.

The better approach is probably to generalize the `DynamicPPL.DebugContext` a bit to be a `CallbackContext`, which allows to just hook into internals of the models to extract the types used. Currently I think `DynamicPPL.check_model_and_trace` captures a bit too many details to be useful for this particular use-case.

Reply:
Do you mind if I just open an issue and leave this for now?

Reply:
Yep!

Reply:
Issue opened, see #2235